internal/provider/resources/block.go

```diff
@@ -116,16 +120,111 @@ func (r *BlockResource) Schema(_ context.Context, _ resource.SchemaRequest, resp
+	// whereas for Read() / Update() we can ask the API for unmasked values to ensure a consistent
+	// state drift check.
+	state.ID = types.StringValue(block.ID.String())
+	state.Created = customtypes.NewTimestampPointerValue(block.Created)
+	state.Updated = customtypes.NewTimestampPointerValue(block.Updated)
+	state.Name = types.StringValue(block.Name)
+	state.TypeSlug = types.StringValue(block.BlockType.Slug)
+
+	return nil
+}
+
 // Create will create the Block resource through the API and insert it into the State.
-//
-//nolint:revive // TODO: remove this comment when method is implemented
 func (r *BlockResource) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
+	var config BlockResourceModel
+	diags := req.Config.Get(ctx, &config)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	blockTypeClient, err := r.client.BlockTypes(config.AccountID.ValueUUID(), config.WorkspaceID.ValueUUID())
+	if err != nil {
+		resp.Diagnostics.Append(helpers.CreateClientErrorDiagnostic("Block Types", err))
+	}
+	blockSchemaClient, err := r.client.BlockSchemas(config.AccountID.ValueUUID(), config.WorkspaceID.ValueUUID())
+	if err != nil {
+		resp.Diagnostics.Append(helpers.CreateClientErrorDiagnostic("Block Schema", err))
+	}
+	blockDocumentClient, err := r.client.BlockDocuments(config.AccountID.ValueUUID(), config.WorkspaceID.ValueUUID())
+	if err != nil {
+		resp.Diagnostics.Append(helpers.CreateClientErrorDiagnostic("Block Document", err))
+	}
+
+	// fetch the BlockType by the provided `type_slug`
+	blockType, err := blockTypeClient.GetBySlug(ctx, config.TypeSlug.ValueString())
+	if err != nil {
+		resp.Diagnostics.Append(helpers.ResourceClientErrorDiagnostic("Block Type", "get_by_slug", err))
+
+		return
+	}
+
+	blockSchemas, err := blockSchemaClient.List(ctx, []uuid.UUID{blockType.ID})
+	if err != nil {
+		resp.Diagnostics.Append(helpers.ResourceClientErrorDiagnostic("Block Schema", "list", err))
+
+		return
+	}
+
+	if len(blockSchemas) == 0 {
+		resp.Diagnostics.AddError(
+			"No block schemas found",
+			fmt.Sprintf("No block schemas found for %s block type slug", config.TypeSlug.ValueString()),
+		)
+
+		return
+	}
+
+	latestBlockSchema := blockSchemas[0]
+
+	// We typed `data` as JSON, as this is the most
+	// flexible way to handle a dynamic schema from the API.
+	// Here, we unmarshal the user-provided `data` JSON string to a map[string]interface{}
+	// because we'll later need to re-marshall the entire BlockDocumentCreate payload
+	// when sending it back up to the API
+	var data map[string]interface{}
+	resp.Diagnostics.Append(config.Data.Unmarshal(&data)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	createdBlockDocument, err := blockDocumentClient.Create(ctx, api.BlockDocumentCreate{
+		Name:          config.Name.ValueString(),
+		Data:          data,
+		BlockSchemaID: latestBlockSchema.ID,
+		BlockTypeID:   latestBlockSchema.BlockTypeID,
+	})
```

Comment:
full disclaimer, i dont love that i did this and im open to suggestions on making this better.  essentially:

1. `data` is a polymorphic JSON from the API
2. we need to type it on the terraform side - T[F does have a concept of dynamic types](https://developer.hashicorp.com/terraform/plugin/framework/handling-data/dynamic-data#considerations), but i opted not to use it to start bc it seemed a little heavy on the caveats. also not sure if it fits our use case here
3. since `data` is defined as JSON in the `.tf` code, we need to pass that forward to the API at some point on creates/updates - but since we already serialize the entire payload, i sort of needed to unmarshal it first to a `map[string]interface` or else it would get double encoded = API will reject it

some things im not sure of yet, but it's worth trying in a followup:
1. can we store the `data` key as a `string` in the golang struct before handing it off to a JSON type in TF? one nice thing is that we use a TF type library that will do some diffs against the normalized JSON, so it will intelligently tell us if there are only whitespace changes, for example
2. any way to exclude marshalling a struct property and pass it forward as-is?

Reply:
Thanks for the context. Thinking this over, too, and I'll post some links I'm still digesting. In the meantime though I'm happy to move forward with this and we can tweak as needed once we get more experience (and testing) in place.

-----

https://discuss.hashicorp.com/t/how-to-handle-json-input-into-data-resource-with-multiple-types-with-the-terraform-sdk/22896/2:

> The typical approach today is, I think, what you already tried: define a string attribute, often with a name ending in _json to be explicit about it, and then decode that value as JSON inside the provider code.

This seems to confirm your `string` attribute thought if I'm reading it correctly.

-----

https://github.com/hashicorp/terraform-plugin-go?tab=readme-ov-file#getting-started:

> When handling requests, [tfprotov5.DynamicValue](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-go/tfprotov5#DynamicValue) types should always have their [Unmarshal](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-go/tfprotov5#DynamicValue.Unmarshal) methods called; their properties should not be inspected directly. The [tftypes.Value](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-go/tfprotov5/tftypes#Value) returned from Unmarshal can be inspected to check [whether it is known](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-go/tfprotov5/tftypes#Value.IsKnown) and subsequently converted to a plain Go type using its [As](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-go/tfprotov5/tftypes#Value.As) method. As will return an error if the Value is not known.

Not sure how much crossover this has with the Terraform Dynamic Type docs you linked earlier, but this was mentioned in the first link I commented.

-----

Small aside, I found https://registry.terraform.io/providers/terraform-provider-graylog/graylog/latest/docs/guides/json-string-attribute so we don't seem to be the only ones doing this 😆

Reply:
Thanks for the context. Thinking this over, too, and I'll post some links I'm still digesting. In the meantime though I'm happy to move forward with this and we can tweak as needed once we get more experience (and testing) in place.

https://discuss.hashicorp.com/t/how-to-handle-json-input-into-data-resource-with-multiple-types-with-the-terraform-sdk/22896/2:

    The typical approach today is, I think, what you already tried: define a string attribute, often with a name ending in _json to be explicit about it, and then decode that value as JSON inside the provider code.

This seems to confirm your string attribute thought if I'm reading it correctly.

https://github.com/hashicorp/terraform-plugin-go?tab=readme-ov-file#getting-started:

    When handling requests, [tfprotov5.DynamicValue](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-go/tfprotov5#DynamicValue) types should always have their [Unmarshal](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-go/tfprotov5#DynamicValue.Unmarshal) methods called; their properties should not be inspected directly. The [tftypes.Value](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-go/tfprotov5/tftypes#Value) returned from Unmarshal can be inspected to check [whether it is known](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-go/tfprotov5/tftypes#Value.IsKnown) and subsequently converted to a plain Go type using its [As](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-go/tfprotov5/tftypes#Value.As) method. As will return an error if the Value is not known.

Not sure how much crossover this has with the Terraform Dynamic Type docs you linked earlier, but this was mentioned in the first link I commented.

Small aside, I found https://registry.terraform.io/providers/terraform-provider-graylog/graylog/latest/docs/guides/json-string-attribute so we don't seem to be the only ones doing this 😆